### PR TITLE
pypechain: code reuse with format_code and write_code

### PR DESCRIPTION
**Pull request by wakamex**

### CODE REVIEW

File: lib/pypechain/pypechain/run_pypechain.py [hunk 1]
```diff
@@ -20,6 +21,19 @@
 from web3.types import ABIFunction
 
 
+def format_and_write_code(path, code):
```
Comment (sentilesdal) [inline lib/pypechain/pypechain/run_pypechain.py:24]: I'd rather this function just be a singular responsibility.  Let's just have it format the code.  Make sure that we are providing types for the parameters and return value and sphynx docs.

Always be weary when you have 'and' in a function name.  sometimes its ok, usually a code smell though.

Leave the writing of the file in the top level for now, its not so bad yet.

**Review by sentilesdal — APPROVED** [review]
this is a great idea, love that we can do this so easily!

File: lib/pypechain/pypechain/run_pypechain.py (hunk 1 above)
Comment (wakamex) [reply lib/pypechain/pypechain/run_pypechain.py:24]: but.. 2 lines of code could be 1 🫣

**Comment by wakamex** [comment]
rebased 🙌

**Review by sentilesdal — APPROVED** [review]

**Comment by sentilesdal** [comment]
please run this on the code first to make sure the codegen isn't garbled before you merge

**Comment by wakamex** [comment]
changes on ERC20Contract.py from main to this PR are all style changes from black:

```diff
1c1
< #pylint: disable=invalid-name
---
> # pylint: disable=invalid-name
4d3
< 
17a17
> 
28a29
> 
39a41
> 
48c50
<     def __call__(self, ) -> "ERC20DecimalsContractFunction":
---
>     def __call__(self) -> "ERC20DecimalsContractFunction":
50a53
> 
59c62,64
<     def __call__(self, spender: str, subtractedValue: int) -> "ERC20DecreaseAllowanceContractFunction":
---
>     def __call__(
>         self, spender: str, subtractedValue: int
>     ) -> "ERC20DecreaseAllowanceContractFunction":
61a67
> 
70c76,78
<     def __call__(self, spender: str, addedValue: int) -> "ERC20IncreaseAllowanceContractFunction":
---
>     def __call__(
>         self, spender: str, addedValue: int
>     ) -> "ERC20IncreaseAllowanceContractFunction":
72a81
> 
81c90
<     def __call__(self, ) -> "ERC20NameContractFunction":
---
>     def __call__(self) -> "ERC20NameContractFunction":
83a93
> 
92c102
<     def __call__(self, ) -> "ERC20SymbolContractFunction":
---
>     def __call__(self) -> "ERC20SymbolContractFunction":
94a105
> 
103c114
<     def __call__(self, ) -> "ERC20TotalSupplyContractFunction":
---
>     def __call__(self) -> "ERC20TotalSupplyContractFunction":
105a117
> 
116a129
> 
125c138,140
<     def __call__(self, _from: str, to: str, amount: int) -> "ERC20TransferFromContractFunction":
---
>     def __call__(
>         self, _from: str, to: str, amount: int
>     ) -> "ERC20TransferFromContractFunction":
127a143
> 
132,133d147
< 
< 
138d151
< 
140d152
< 
142d153
< 
144d154
< 
146d155
< 
148d156
< 
150d157
< 
152d158
< 
154d159
< 
156d160
< 
167,168c171,175
<         self.address = address if address else cast(ChecksumAddress, "0x0000000000000000000000000000000000000000")
< 
---
>         self.address = (
>             address
>             if address
>             else cast(ChecksumAddress, "0x0000000000000000000000000000000000000000")
>         )
172d178
< 
178,179c184
< 
<     functions: ERC20ContractFunctions
\ No newline at end of file
---
>     functions: ERC20ContractFunctions
```

**Pull request merged**